python/cudf/dataframe/string.py

```diff
@@ -590,6 +592,64 @@ def to_array(self, fillna=None):
 
         return self.to_arrow().to_pandas()
 
+    def serialize(self, serialize):
+        header = {
+            'null_count': self._null_count,
+        }
+        frames = []
+        sub_headers = []
+        if self.data.size() > 0:
+            arr = np.arange(self.data.size(),dtype=np.int32)
+            d_arr = rmm.to_device(arr)
+            self.data.len(d_arr.device_ctypes_pointer.value)
+            nchars = sum(d_arr.copy_to_host()) + self.null_count * 2
+            values = np.empty(nchars, dtype=np.int8)
+            offsets = np.empty(nchars+1, dtype=np.int32)
+            self.data.to_offsets(values, offsets)
+        else:
+            values, offsets = None, None
+        for i, item in enumerate([self.mask, values, offsets]):
+            if i == 0:
+                sheader, [frame] = serialize(item)
+            else:
+                item_ndarray = cuda.to_device(item)
+                sheader, [frame] = serialize(item_ndarray)
+            sub_headers.append(sheader)
+            frames.append(frame)
+
+        header["nvstrings"] = len(self.data)
+        header["subheaders"] = sub_headers
+        return header, frames
```

Comment:
We should be able to do this entirely on the device using a combination of rmm and our `cudf.utils.cudautils`. If you follow the logic in the `to_arrow` method and replace the numpy calls with `cudautils` calls, we should be able to just pass an extra parameter to the `to_offsets` call of `bdevmem=True` to instruct it to use the device memory.